### PR TITLE
Project hygiene: Reference and Generator projects

### PR DESCRIPTION
**Pull request by ViktorHofer**
Part of https://github.com/dotnet/runtime/pull/67532

### [Infrastructure changes](https://github.com/dotnet/runtime/commit/36cbe0b620a271273ea3abae2116eaa5f34c3c26) 

1. Cleaning-up the generators.targets file by simplifying conditions.
2. Enable "default items" via the `<EnableDefaultItems />` SDK switch
   for anything except source and test projects.
3. Remove outdated infrastructure that isn't needed anymore based on
   project file changes.
4. Avoid duplicate resource imports for OOB projects.

### [Ref project file hygiene](https://github.com/dotnet/runtime/commit/e88621db71ec916c0f16a77f01a7165c612928ea) 

1. Use the SDK's default items feature so that compile items in the
   project directory are imported automatically.

   Changes made, mimic the usage in the Microsoft.Extensions.* projects
   which already used the default items feature.
   I made sure that the compile items included before and after are
   identical.

2. Remove the unnecessary declaration of `<Reference />` and
   `<ProjectReference />` items in out-of-band projects as they already
   build on top of the shared framework (targeting pack) and make that
   the default behavior for anything that is out-of-band.

3. Clean-up Reference and ProjectReference items that weren't used by
   the compiler anymore, i.e. System.Runtime.ComilerServices.Unsafe and
   System.Runtime.Extensions (as those are now facades).

4. Add empty lines to separate property and item groups from each other
   so that project files are easier to read and look the same as
   project files that "dotnet new" produces.

### [Generator project file hygiene](https://github.com/dotnet/runtime/commit/fd8b1b7123d01fa4778e02082a48f11d1cfc9eca)
Same changes as for the ref project file hygiene.

### CODE REVIEW

File: src/coreclr/nativeaot/Directory.Build.props [hunk 1]
```diff
@@ -6,7 +6,7 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <EnableDefaultItems>false</EnableDefaultItems>
+    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
     <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
```
Comment (bartonjs) [inline src/coreclr/nativeaot/Directory.Build.props:10]: Why?

EnableDefaultItems being off gives a much better error when things get out of state, and it being on leads to confusion when some particles get created (e.g. `copy Foo.cs Foo.bak.cs`... now the project fails)

I really really really (etc) do not want us to ever turn this on.

So: Why turn it on?

File: src/libraries/Directory.Build.targets [hunk 2]
```diff
@@ -48,14 +48,13 @@
...
-                                                   $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '$(NETCoreAppCurrentVersion)')) and
+    <DisableImplicitFrameworkReferences Condition="'$(TargetFrameworkVersion)' == 'v$(NetCoreAppCurrentVersion)' and
                                                    ('$(IsNETCoreAppRef)' == 'true' or '$(IsNETCoreAppSrc)' == 'true')">true</DisableImplicitFrameworkReferences>
-    <!-- Disable implicit assembly references for .NETCoreApp refs and sources. -->
+    <!-- Disable implicit assembly references for .NETCoreApp sources. -->
     <DisableImplicitAssemblyReferences Condition="'$(DisableImplicitAssemblyReferences)' == '' and
                                                   '$(DisableImplicitFrameworkReferences)' != 'true' and
                                                   '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
-                                                  ('$(IsReferenceAssemblyProject)' == 'true' or '$(IsSourceProject)' == 'true')">true</DisableImplicitAssemblyReferences>
+                                                  '$(IsSourceProject)' == 'true'">true</DisableImplicitAssemblyReferences>
```
Comment (bartonjs) [inline src/libraries/Directory.Build.targets:57]: Again... why?

I was under the impression that we liked the explicitness in our references, even though they occasionally cause frustration.

(This one I don't really care about, but I don't understand the why)

**Comment by bartonjs** [comment]
> I made sure that the compile items included before and after are
identical.

The best way to to do that is to leave them as explicitly required :smile:.

If the inconsistency we're upset about is "extensions use default items and none of the rest of our projects do" then make extensions match the rest of the tree, not the other way around.

**Review by bartonjs — CHANGES_REQUESTED** [review]
Since there's no tracking issue for this (it appears to be work out of the blue) the change to use DefaultItems should not be merged.

Removing stale references and normalizing whitespace are reasonable hygiene changes... but changing how we structure our project files altogether isn't.

**Comment by ViktorHofer** [comment]
@bartonjs for context, I opened https://github.com/dotnet/runtime/pull/67420 last week to see what area owners would think of it. [@stephentoub commented on the PR](https://github.com/dotnet/runtime/pull/67420#discussion_r840642106) and I read it as he was asking why we can't do this for all projects and not just for System.Data.Common. Now that I'm reading his comment again I'm unsure if I did misinterpret his question but it would be best if Stephen could chime in and clarify what he meant.

Aside from that, I spoke with him offline about this change today and I didn't hear any strong objections yet but it could be that he hasn't yet read my commit summaries in the PRs.

**Comment by stephentoub** [comment]
> Now that I'm reading his comment again I'm unsure if I did misinterpret his question but it would be best if Stephen could chime in and clarify what he meant.

I didn't mean to imply that.  My comment was more "why are we doing this for System.Data.Common.  We don't do it for anything else."  I don't see a good reason to do it for any of our projects.

>  I spoke with him offline about this change today and I didn't hear any strong objections yet but it could be that he hasn't yet read my commit summaries in the PRs.

Yes, I hadn't looked at the PR yet beyond seeing that you were modifying 500 files.  You were asking how to get it reviewed quickly, and I suggested it was way too big to expect a quick turnaround, and that small PRs would be better.

**Comment by ViktorHofer** [comment]
Closing this PR. Please see https://github.com/dotnet/runtime/pull/67532#issuecomment-1090464863 for more details.

**Pull request closed without merging**